File: jotupix_demo_ios/JotupixDemo/Jotupix/Core/JProtocol.m
// JProtocol.m
#import "JProtocol.h"
#import "JLog.h"

#define TAG     "JProtocol"

// Protocol control characters
#define PROTOCOL_START_CHAR           0x01    // Start marker
#define PROTOCOL_ESC_CHAR             0x02    // Escape marker
#define PROTOCOL_ESC_XOR_CHAR         0x04    // Escape XOR mask
#define PROTOCOL_END_CHAR             0x03    // End marker

#define PROTOCOL_SEND_MAX_LEN         65535   // Max send length

typedef NS_ENUM(uint8_t, PROTOCOL_RECV_STATUS_E)
{
    E_PROTOCOL_RECV_STATUS_DATA = 0,      // Currently receiving actual data bytes
    E_PROTOCOL_RECV_STATUS_START,         // Waiting for next frame start
    E_PROTOCOL_RECV_STATUS_DATALEN_H,     // Receiving data length high byte
    E_PROTOCOL_RECV_STATUS_DATALEN_L      // Receiving data length low byte
};

@interface JProtocol () {
    // tx buffer and sender
    id<IJSend> _sender;
    uint8_t _txBuffer[JPROTOCOL_TX_BUFF_SIZE];

    // rx state variables
    id<IJParseCallback> _callback;
    uint8_t _parseBuffer[JPROTOCOL_RX_PARSE_BUFF_SIZE];
    uint8_t _recvStatus;
    uint16_t _recvIndex;
    uint16_t _parseDataLen;
    BOOL _inEsc;
}
@end

@implementation JProtocol

#pragma mark - Init

/// Initialize protocol with sender and callback
- (void)init:(id<IJSend>)sender callback:(id<IJParseCallback>)callback {
    _sender = sender;
    _callback = callback;
}

#pragma mark - Send

/// Pack and send data using start/length/escape/end protocol
- (int)send:(const uint8_t *)pu8Data length:(uint32_t)len {
    if (!_sender) {
        JLogE(TAG, "Send data fail, send is NULL\r\n");
        return -1;
    }
    if (!pu8Data || len == 0 || len > PROTOCOL_SEND_MAX_LEN) {
        JLogE(TAG, "Send data fail, input error\r\n");
        return -1;
    }

    uint16_t index = 0;

    _txBuffer[index++] = PROTOCOL_START_CHAR;  // Start marker

    // High byte of length
    uint8_t tmp = (len >> 8) & 0xFF;
    [self xorAndAppend:tmp index:&index];

    // Low byte of length
    tmp = len & 0xFF;
    [self xorAndAppend:tmp index:&index];

    // Append data with escape processing
    for (uint32_t i = 0; i < len; i++) {
        [self xorAndAppend:pu8Data[i] index:&index];

        // Flush partial buffer if near full
        if (index > JPROTOCOL_TX_BUFF_SIZE - 2) {
            int ret = [_sender send:_txBuffer length:index];
            if (ret != 0) return -1;
            index = 0;
        }
    }

    _txBuffer[index++] = PROTOCOL_END_CHAR;   // End marker

    return [_sender send:_txBuffer length:index];
}

/// Escape byte handler
- (void)xorAndAppend:(uint8_t)value index:(uint16_t *)index {
    if (value > 0 && value < PROTOCOL_ESC_XOR_CHAR) {
        _txBuffer[(*index)++] = PROTOCOL_ESC_CHAR;            // escape prefix
        _txBuffer[(*index)++] = (value | PROTOCOL_ESC_XOR_CHAR); // escaped value
    } else {
        _txBuffer[(*index)++] = value;
    }
}

#pragma mark - Parse

/// Parse incoming raw data stream using protocol state machine
- (void)parse:(const uint8_t *)pu8Data length:(uint32_t)len {
    if (!pu8Data || len == 0) {
        JLogE(TAG, "Parse data fail, input error\r\n");
        return;
    }

    for (uint32_t i = 0; i < len; i++) {
        uint8_t byte = pu8Data[i];

        switch (byte) {

            case PROTOCOL_START_CHAR:
                // Reset state for a new frame
                _parseDataLen = 0;
                _recvIndex = 0;
                _inEsc = NO;
                _recvStatus = E_PROTOCOL_RECV_STATUS_DATALEN_H;
                break;

            case PROTOCOL_ESC_CHAR:
                _inEsc = YES;   // Next byte is escaped
                break;

            case PROTOCOL_END_CHAR:
                // End of frame → validate + callback
                if (_recvStatus == E_PROTOCOL_RECV_STATUS_DATA &&
                    _parseDataLen > 0 &&
                    _parseDataLen <= JPROTOCOL_RX_PARSE_BUFF_SIZE) {

                    if (_callback) {
                        // Notify upper layer
                        [_callback onParseComplete:_parseBuffer length:_parseDataLen];
                    }
                }
                _recvStatus = E_PROTOCOL_RECV_STATUS_START;
                break;

            default:
                // Handle escape transformation
                if (_inEsc) {
                    byte ^= PROTOCOL_ESC_XOR_CHAR; // XOR unescape 
                    _inEsc = NO;
                }

                switch (_recvStatus) {

                    case E_PROTOCOL_RECV_STATUS_DATA:
                        // Collect payload bytes
                        if (_recvIndex < _parseDataLen) {
                            _parseBuffer[_recvIndex++] = byte;
                        }
                        break;

                    case E_PROTOCOL_RECV_STATUS_DATALEN_H:
                        // Store high length byte
                        _parseDataLen = (byte << 8);
                        _recvStatus = E_PROTOCOL_RECV_STATUS_DATALEN_L;
                        break;

                    case E_PROTOCOL_RECV_STATUS_DATALEN_L:
                        // Store low length byte + validate range
                        _parseDataLen |= byte;

                        if (_parseDataLen == 0 ||
                            _parseDataLen > JPROTOCOL_RX_PARSE_BUFF_SIZE) {
                            // Invalid length, reset
                            _parseDataLen = 0;
                            _recvStatus = E_PROTOCOL_RECV_STATUS_START;
                        } else {
                            // Ready to receive payload
                            _recvIndex = 0;
                            _recvStatus = E_PROTOCOL_RECV_STATUS_DATA;
                        }
                        break;

                    default:
                        break;
                }
                break;
        }
    }
}

@end

File: jotupix_demo_windows/jotupix/core/JProtocol.cpp
//#include "pch.h"
#include "JProtocol.h"
#include "JLog.h"

#define TAG     "JProtocol"

#define PROTOCOL_START_CHAR           0x01      // Start mark
#define PROTOCOL_ESC_CHAR             0x02      // Protocol Flags
#define PROTOCOL_ESC_XOR_CHAR         0x04      // Protocol XOR value
#define PROTOCOL_END_CHAR             0x03      // End Marker

#define PROTOCOL_SEND_MAX_LEN         65535

typedef enum _PROTOCOL_RECV_STATUS_E
{
    E_PROTOCOL_RECV_STATUS_DATA = 0, // Receiving Data
    E_PROTOCOL_RECV_STATUS_START,    // Receive start signal
    E_PROTOCOL_RECV_STATUS_DATALEN_H, // Receive data length high bit
    E_PROTOCOL_RECV_STATUS_DATALEN_L, // Receive data length low bit
} PROTOCOL_RECV_STATUS_E;

#define iXor(V) \
    if ((V) > 0 && (V) < PROTOCOL_ESC_XOR_CHAR) \
    { \
        m_au8TxBuffer[u16Index++] = PROTOCOL_ESC_CHAR; \
        m_au8TxBuffer[u16Index++] = ((V) | PROTOCOL_ESC_XOR_CHAR); \
    } \
    else \
    { \
        m_au8TxBuffer[u16Index++] = (V); \
    }

void JProtocol::Init(IJSend* pfnSender, IJParseCallback *pfnCallback)
{
    m_pfnSender = pfnSender;
    m_pfnCallback = pfnCallback;
}

int JProtocol::Send(const uint8_t* pu8Data, uint32_t u32Len)
{
    uint32_t i = 0;
    uint16_t u16Index = 0;
    uint8_t u8Tmp = 0;
    int nRet = -1;

    if (m_pfnSender == NULL)
    {
        JLogE(TAG, "Send data fail, send is NULL\r\n");
        return -1;
    }

    if (pu8Data == NULL || u32Len == 0 || u32Len > PROTOCOL_SEND_MAX_LEN)
    {
        JLogE(TAG, "Send data fail, input error\r\n");
        return -1;
    }

    m_au8TxBuffer[u16Index++] = PROTOCOL_START_CHAR;

    // send data length
    u8Tmp = (u32Len >> 8) & 0xFF;
    iXor(u8Tmp);

    u8Tmp = u32Len & 0xFF;
    iXor(u8Tmp);

    for (i = 0; i < u32Len; i++)
    {
        iXor(pu8Data[i]);

        if (u16Index > JPROTOCOL_TX_BUFF_SIZE - 2)
        {
            nRet = m_pfnSender->Send(m_au8TxBuffer, u16Index);
            if (nRet != 0)
            {
                return -1;
            }

            u16Index = 0;
        }
    }

    m_au8TxBuffer[u16Index++] = (PROTOCOL_END_CHAR);

    return m_pfnSender->Send(m_au8TxBuffer, u16Index);
}

void JProtocol::Parse(const uint8_t* pu8Data, uint32_t u32Len)
{
    uint32_t i = 0;
    uint8_t u8Byte = 0;

    if (pu8Data == NULL || u32Len == 0)
    {
        JLogE(TAG, "Parse data fail, input error\r\n");
        return;
    }

    for (i = 0; i < u32Len; i++)
    {
        u8Byte = pu8Data[i];

        switch (u8Byte)
        {
        case PROTOCOL_START_CHAR:      // Start Marker
            m_u16ParseDataLen = 0;
            m_bInEsc = false;
            m_u8RecvStatus = E_PROTOCOL_RECV_STATUS_DATALEN_H;  // Waiting to receive data length
            break;
        case PROTOCOL_ESC_CHAR:     //0x02
            m_bInEsc = true;
            break;
        case PROTOCOL_END_CHAR:    // End Marker
            if (m_u8RecvStatus == E_PROTOCOL_RECV_STATUS_DATA
                && m_u16ParseDataLen <= JPROTOCOL_RX_PARSE_BUFF_SIZE
                && m_u16ParseDataLen > 0)
            {
                if (m_pfnCallback != NULL)
                {
                    m_pfnCallback->onParseComplete(m_au8ParseBuffer, m_u16ParseDataLen);
                }
            }

            m_u8RecvStatus = E_PROTOCOL_RECV_STATUS_START; // Waiting for receiving start flag
            break;
        default:
            if (m_bInEsc)   //0x02
            {
                u8Byte ^= PROTOCOL_ESC_XOR_CHAR;   // XOR operation
                m_bInEsc = false;
            }

            switch (m_u8RecvStatus)
            {
            case E_PROTOCOL_RECV_STATUS_DATA:
            {
                if (m_u16RecvIndex < m_u16ParseDataLen)
                {
                    m_au8ParseBuffer[m_u16RecvIndex] = u8Byte;
                    m_u16RecvIndex++;
                }
            }
            break;
            case E_PROTOCOL_RECV_STATUS_DATALEN_H:
            {
                m_u16ParseDataLen = u8Byte << 8;
                m_u8RecvStatus = E_PROTOCOL_RECV_STATUS_DATALEN_L;
            }
            break;
            case E_PROTOCOL_RECV_STATUS_DATALEN_L:
            {
                m_u16ParseDataLen |= u8Byte;

                if (m_u16ParseDataLen > JPROTOCOL_RX_PARSE_BUFF_SIZE || m_u16ParseDataLen == 0)
                {
                    m_u8RecvStatus = E_PROTOCOL_RECV_STATUS_START;
                    m_u16ParseDataLen = 0;
                }
                else
                {
                    m_u8RecvStatus = E_PROTOCOL_RECV_STATUS_DATA;
                    m_u16RecvIndex = 0;
                }
            }
            break;
            default:
                break;
            }
            break;
        }
    }
}